### DISCUSSION TITLE
404 response code: not able to connect to the asyncio server deployed in testing environment

### DISCUSSION THREAD

**Discussion by nsrawat009** [miguelgrinberg/python-socketio#1339] [Q&A]
I have deployed asyncio server in one of the testing environment of my firm successfully but iam not able to make connection to the server with Postman or python client although i was able to connect and get response successfully in my local testing.I dont know it is the issue with routing or incorrect url.
Iam unsure about the socketio_path arguement in ASGI app,whether i have to provide it or not since bydefault it takes socket.io path whether it should I define it inside sio_app or not??

```python
#intializing socketio server
sio = socketio.AsyncServer(
    async_mode='asgi',
    cors_allowed_origins=['*'],
    ping_timeout=6000,   
    ping_interval=12000,
    logger=True,
    engineio_logger=True
)

sio_app = socketio.ASGIApp(
    socketio_server=sio,
    other_asgi_app=app,
    static_files = {
    '/static': './public'
}
)
```

Note:here app is another fast api app dedicated for rest endpoints

```python
#Event handler to connect client to socketio server
@sio.event()
async def connect(sid, environ):
    logger.info('user trying to connect')
    username = environ.get('HTTP_X_USERNAME')
    logger.info(f"user trying to connect: {username}")
    if not username:
        return False
    async with sio.session(sid) as session:
        session['username'] = username
    await sio.emit(event='user_joined',data=username)



@sio.on('recommend_gsi',namespace='/paragraph_grammar/gsi_suggestions')
async def gsi_recommendation(sid: str, data: Dict[str, Any]):
    logger.info("GSI Recommendations through socketio API")
    # Extract authorization and accept-language headers from data
    try:
        # Extract authorization and accept-language headers from data
        gsi_collection_headers = {
            "Authorization": data.get("Authorization"),
            "accept-language": data.get("accept-language", "en").lower()
        }
```
********************************

is my url correct??: (ws://nebula.nslhub.com/dsd-orch/paragraph_grammar/gsi_suggestions
Note: nebula.nslhub.com is the dns name of the environment
dsd-orch is the security layer
/paragraph_grammar/gsi_suggestions is the namespace  of the event
Also i have not implemented ssl configuration yet.
 
Note in my microservice the request comes via istio -> kubernetes pod -> microservice -> end point

Currently iam getting 404 response code I will try removing static files which i have currently provided.What else should i check,Kindly support iam stuck

**Comment by miguelgrinberg** [comment]
Socket.IO is a bit strange in how the URL is specified.

The connection URL should be your hostname followed by your namespace. So try using `ws://nebula.nslhub.com/paragraph_grammar/gsi_suggestions`.

The path to the Socket.IO endpoint is given in the `path` argument, which I think in your case should be `/dsd-orch/socket.io`.

If this does not work, you should probably try to make your HTTP endpoints work first, then adapt your url and path accordingly.

**Reply by nsrawat009** [reply]
HTTP endpoints are working fine,actually entire service is running on REST based endpoints and infra is well built for REST based endpoints.this is the first  time we are trying to integrate sockets into our microservice.
https://nebula.nslhub.com/dsd-orch/paragraph_grammar/extract_bpmn is one of the REST based endpoint
I will try to change socketio path to /dsd-orch/socket.io although tried previously only with /dsd-orch but was not working

**Reply by nsrawat009** [reply]
![Uploading dsd-orch 2.png…]()

![Uploading dsd-orch 2.png…]()
Hi @miguelgrinberg /dsd-orch/socket.io didn't worked. When debugging logs through grafana i can see in the pods the request is going from istio service>>dsd-orhcestrator service>>GET dsd-orch/socket.io. However it is supposed to go like istio service>>paragraph grammar service>>socket event via dsd layer ofcourse similar to REST request. What are we missing here?? Are the issues in our side or network routing ?? How can we make this work?? I have POC ready in my local biut facing issues while deploying.
Is python-socketio compatible with deployment via gunicorn and fastapi?

I have attached REST vs Socket comparison images.
![rest1](https://github.com/miguelgrinberg/python-socketio/assets/68178464/9cfcedc2-2c1b-4bbd-8f3a-5802f9d10f56)

![dsd-orch](https://github.com/miguelgrinberg/python-socketio/assets/68178464/55e4c999-24a8-4724-9db3-2609a77bdac0)

**Reply by miguelgrinberg** [reply]
> Is python-socketio compatible with deployment via gunicorn and fastapi?

It should if you do it correctly. For async apps I use Uvicorn, but Gunicorn with the Uvicorn worker should work as well. You can try this outside of K8s if you need to confirm.

As I explained above, you need to find the right combination of settings for `url` and `path`. You are showing me screenshots, which is not very useful. What would be useful is to know what's the value of `path` in the Socket.IO server, and what URL(s) the client browser sends for regular requests and Socket.IO requests.

**Reply by nsrawat009** [reply]
sio_app = socketio.ASGIApp(
    socketio_server=sio,
    socketio_path="/dsd-orch/socket.io"
    other_asgi_app=app,
    static_files = {
    '/static': './public'
}
)
  I kept socketio_path="/dsd-orch/socket.io" as you suggested before and sample url for REST requests is https://nebula.nslhub.com/dsd-orch/paragraph_grammar/extract_bpmn
Here:  paragraph_grammar is the microservice deployed in kubernetes pods and extract_bpmn is the HTTP endpoint
 whereas in the socketio case i tried  to hit 
 ws://nebula.nslhub.com/dsd-orch/socket.io/paragraph_grammar/gsi_suggestions
 ws://nebula.nslhub.com/dsd-orch/paragraph_grammar/gsi_suggestions
 ws://nebula.nslhub.com/paragraph_grammar/gsi_suggestions
 Here :paragraph_grammar/gsi_suggestions  is the namespace
 but all are failing to connect with Postman(websocket and socketio both tried) and python client iam getting folowing error in python client
 User
Attempting WebSocket connection to ws://nebula.nslhub.com/dsd-orch/socket.io/?transport=websocket&EIO=4
Connection error
Failed to connect: Connection error
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x7377b861cd00>
Unclosed connector
connections: ['[(<aiohttp.client_proto.ResponseHandler object at 0x7377b860d720>, 6060.1061981)]', '[(<aiohttp.client_proto.ResponseHandler object at 0x7377b860e4a0>, 6060.2374707)]']
connector: <aiohttp.connector.TCPConnector object at 0x7377b861cdc0>

**Reply by miguelgrinberg** [reply]
Any difference if you remove `path` in the server, but keep it in the client?

**Reply by nsrawat009** [reply]
After doing cahnges in dsd-orch repo my request is reaching the service now but getting GET 404 not found error not sure its the socketio server throwing that error or the fastapi server how can I verify that and how to resolve that.

**Reply by nsrawat009** [reply]
not mentioning path in the server is not working at all when i kept /dsd-orch/socket.io in path arguement atleast request is reaching the pod

**Reply by miguelgrinberg** [reply]
Unfortunately I'm not sure how I can help more with this. You see that this is clearly a routing problem. If the requests reach the Python process, then it is a matter of looking at what the URL that is received in Python looks like, and then setting the `path` argument in the server to make sure the URL is a match. That's really all there is to it, but you'll need to add some logging in the `ASGIApp` class to see the URL.

In case it helps, here is the place where the Socket.IO routing logging happens (in the python-engineio package): https://github.com/miguelgrinberg/python-engineio/blob/main/src/engineio/async_drivers/asgi.py#L64-L67. Once you know what URL arrives at the Python server, you just need to set `socketio_path` (called `engineio_path` in the lower level code I linked) so that the condition in this if-statement is true, which will forward the request to the Socket.IO logic.

**Reply by nsrawat009** [reply]
I was able to connect to the socketio server by changing the socketio_path to "/paragraph_grammar/websocket".
Below code worked for me:
#intializing socketio server
sio = socketio.AsyncServer(
    async_mode='asgi',
    cors_allowed_origins=[],
    always_connect=True,    
    ping_timeout=6000,   
    ping_interval=12000,
    logger=True,
    engineio_logger=True
)

# Define the functions to handle startup and shutdown events for socketio APP intialization
async def put_startup_logs():
    # Put logs for connection establishment
    logger.info("Socket.IO connection established.")  

async def put_shutdown_logs():
    # Put logs for connection termination
    logger.info("Socket.IO connection terminated.")  



# Create ASGI application for Socket.IO
sio_app = socketio.ASGIApp(
    socketio_server=sio,
    other_asgi_app=app,
    on_startup=put_startup_logs,  # Log when the connection is established 
    on_shutdown=put_shutdown_logs,  # Log when the connection is ended 
    socketio_path="/paragraph_grammar/websocket"    
)
Moreover, i was able to connect with wss not ws since our environments are secured.
"wss://nebula.nslhub.com/dsd-orch/paragraph_grammar/websocket/?transport=websocket&EIO=4" worked for me.
Also iam able to emit events and get the responses

/dsd-orch/socket.io  path was transferring the call to one of our service DSD and we were getting authentification fail previously.
